common/achievement/get.ts

```diff
@@ -0,0 +1,67 @@
+import { prisma } from '../prisma';
+import { Achievement_template, User_achievement, achievement_action_type_enum, achievement_type_enum } from '../../graphql/generated';
+import { Step, Achievement, achievement_state } from '../../graphql/types/achievement';
+import { User } from '../user';
+
+const getUserAchievements = async (user: User): Promise<Achievement[]> => {
+    const userAchievements = await prisma.user_achievement.findMany({
+        where: { userId: user.userID },
+        include: { template: true },
+    });
+    const userAchievementGroups: { [group: string]: User_achievement[] } = {};
+    userAchievements.forEach((ua) => {
+        if (!userAchievementGroups[ua.template.group]) {
+            userAchievementGroups[ua.template.group] = [];
+        }
+        userAchievementGroups[ua.template.group].push(ua);
+    });
+    const achievements: Achievement[] = createFrontendAchievements(userAchievementGroups);
+    return achievements;
+};
+
+const createFrontendAchievements = (groups: { [group: string]: User_achievement[] }): Achievement[] => {
+    const achievements: Achievement[] = [];
+    for (const group in groups) {
+        const sortedGroupAchievements = groups[group].sort((a, b) => a.groupOrder - b.groupOrder);
+        const currentAchievementIndex = sortedGroupAchievements.findIndex((ua) => !ua.achievedAt) - 1;
```

Comment:
Can't this cause an out of index error if the first achievement in the list (index: 0) isn't achieved yet?

Reply:
in this line we are searching for the first index to not have an achievedAt value. So this will not cause an error. But we would run in a different issue. If there was found no element the result would be -1. Also we have to return the index without substracting 1 so we get the real index value.

I added a resultIndex variable that points to null if index < 0 and replaced all currentAchievementIndexes with this variable.